syne_tune/optimizer/schedulers/searchers/searcher.py

```diff
@@ -595,15 +658,43 @@ def _generate_remaining_candidates(self) -> List[dict]:
         return remaining_candidates
 
     def _generate_all_candidates_on_grid(self) -> List[dict]:
-        # Get hp values that are specified as Domain
+        """
+        Generates all configurations to be evaluated by placing a regular, equally spaced grid over the configuration space
+        Returns:
+            The set of all configurations on grid.
+        """
         hp_keys = []
         hp_values = []
-        for key, hp_range in reversed(list(self.config_space.items())):
-            if isinstance(hp_range, Domain):
-                hp_keys.append(key)
+        ## adding categorical, finiteRange, scalar parameters
+        for hp, hp_range in reversed(list(self.config_space.items())):
+            if isinstance(hp_range, Float) or isinstance(hp_range, Integer):
+                continue
+            if isinstance(hp_range, Categorical):
+                hp_keys.append(hp)
                 hp_values.append(hp_range.categories)
+            elif isinstance(hp_range, FiniteRange):
+                hp_keys.append(hp)
+                hp_values.append(hp_range.values)
+            elif not isinstance(hp_range, Domain):
+                hp_keys.append(hp)
+                hp_values.append([hp_range])
+
+        ## adding float, integer parameters
+        for hpr in self._hp_ranges._hp_ranges:
+            if hpr.name not in hp_keys:
+                _hpr_nsamples = self.num_samples[hpr.name]
+                _normalized_points = [
+                    (idx + 0.5) / _hpr_nsamples for idx in range(_hpr_nsamples)
+                ]
+                _hpr_points = [
+                    hpr.from_ndarray(np.array([point])) for point in _normalized_points
+                ]
+                _hpr_points = list(set(_hpr_points))
+                hp_keys.append(hpr.name)
+                hp_values.append(_hpr_points)
 
         hp_values_combinations = list(product(*hp_values))
+
         if self._shuffle_config:
             shuffle(hp_values_combinations)
 
```

Comment:
Sorry, Mina. While this is no your code, I'd love this to get fixed (don't know how this got through our reviews).

Two things:
- Please store hp_values_combinations (list of tuples) and hp_keys as members of the class, and not all_candidates_on_grid (list of dict)
- Create the dict on demand in _next_candidate_on_grid
- Do not use pop(0) in _next_candidate_on_grid. Leave the list of all tuples constant, and use a position index

Using pop is potentially really expensive on a large list. And it is totally unnecessary here. My favourite would be not to create this list of tuples at all, but let us keep this.

Reply:
I addressed first and third of bullet points above. But about the second bullet: We need to create the dict once in _generate_all_candidates_on_grid() so that we remove duplicates (if points_to_evaluate is not None, duplicates will exist).

Reply:
That is not so important. Strictly speaking, a scheduler may return a config more than once, it is just we try to avoid that. In this case, it makes things complicated.

Reply:
You can just do `self.random_state.shuffle(...)`